### PR TITLE
Project: Private events

### PR DESCRIPTION
**Pull request by t-yanick**
In this project, we build a site similar to `Everbrite` which:
- Allow users to create events and
- Manage their signups

We made use of Session to authenticate users.

Please kindly provide feedback on how we can improve this project.
Thanks

### CODE REVIEW

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Great effort with this project, team! 👍 
You have done an amazing job so far! 👏 
However, we still need to implement some changes in order to get the project approved 💪 
Kindly, check my comments below 👇  apply the changes and resubmit your project for a new code review ⏩ 
Good luck! 😺 
## Mandatory changes
- Events controller
   * [x] lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
    * [x] lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- Users controller, show action:
    * [x] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  * [x] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  * [x] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- Events model: kindly add an index for creator_id on your DB. Since the foreign key columns will be used often to retrieve related information, it's important to index them.
More details:

   * https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html
   * https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys.
- Kindly add tests with RSpec to your project:
  * [ ] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  * [ ] authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  * [ ] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- Your app is throwing an error when clicking the link to create an event after the sign up, kindly fix it.
`undefined method `created_events' for nil:NilClass`
- Also an error is reported when creating an event `ActionView::SyntaxErrorInTemplate in EventsController#show`
- The events show views also reports a problem `ActionView::SyntaxErrorInTemplate in EventsController#show`
- Kindly create a personalized README file different to the one provided by Rails.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 

### Hello team 😀,

**Great job implementing the changes 💪. However, there are still a few fixes required.**

1). **In the website homepage, add a link to help navigate through the logged-in user's profile page instead of typing it manually.**

2). **Remove the new event link/button at the bottom of your homepage if no user is logged in.**

3). **Your integration tests should be stored in a folder called spec. You have stored yours in a folder called test and so when I run `bundle exec rspec` or `rspec`, I get no result. The test command you provided in your readme instruction manual throws the following error:**
rspec

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

**Comment by t-yanick** [comment]
Hi ,

Thanks for your review and feedback. We have

- Added a link to the homepage to navigate when a user is logged-in.
- Removed the Event link/button when no user is logged-in.
- We have put all integrations test in the `spec` folder.

Waiting for your feedback

File: app/models/attendance.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Attendance < ApplicationRecord
+  belongs_to :attendee, class_name: 'User'
+  belongs_to :attended_event, class_name: 'Event'
+end
```
Comment (ghost) [inline app/models/attendance.rb:4]: Add tests for the above associations in `spec/models/attendance_spec.rb`.

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_many :created_events, foreign_key: :creator_id, class_name: 'Event'
+  has_many :attendances, foreign_key: :attendee_id
+  has_many :attended_events, through: :attendances
```
Comment (ghost) [inline app/models/user.rb:4]: Add tests for the above associations in `spec/models/user_spec.rb`.

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class Event < ApplicationRecord
+  belongs_to :creator, foreign_key: :creator_id, class_name: 'User'
+  has_many :attendances, foreign_key: :attended_event_id
+  has_many :attendees, through: :attendances
```
Comment (ghost) [inline app/models/event.rb:4]: Add tests for this in the`spec/models/event_spec.rb`.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team 😄,

**Great job implementing the changes 💪. However, there are still a few fixes needed.**

1). **Write integration tests for your views as well. Remember to create them in the spec directory. Ensure you don't create a spec file and leave it empty. Example, your `attendance_spec.rb` is empty.**

2). **Fix the rubocop errors.**
rubocop


3). **Check for more in the comments below.**

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

File: spec/features/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      expect(page).to have_content("Name can't be blank")
+      expect(page).to have_content("Username can't be blank")
+    end
+    it 'create a new user' do
+      visit signup_path
+      fill_in 'Name', with: 'w'
+      fill_in 'Username', with: 'w1'
+      click_on 'Create User'
+      expect(page).to have_content('User was successfully created.')
+    end
```
Comment (addod19) [inline spec/features/user_spec.rb:19]: Our test is failing, let us kindly fix it

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @t-yanick and @mohamedSabry0, congratulations on this project :+1: Great job fixing all the comments outlined by the previous TSE's :slightly_smiling_face: However, we have one issue that needs fixing. Kindly check the comment below:
- We have one failing test, which is the features/user_spec.rb.The test is unable to create a new user as outlined in the comment below

Once this is fixed, kindly submit another code review :smile: 
_Happy coding_  :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Comment by mohamedSabry0** [comment]
@addod19 Thank you for your feedback. Can we discuss it somehow? as I'm not able to reproduce the error.

**Comment by mohamedSabry0** [comment]
Thank you all for your great feedbacks. I hope this time all tests will pass. also, if you have **chrome** installed you can use it by replacing line 44 in `rails_helper.rb` with `Capybara.default_driver = :selenium_chrome`.

tests_result

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️
### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**